### PR TITLE
feat(sltp): enable notifications for sl/tp orders created through modal

### PR DESCRIPTION
**Pull request by moo-onthelawn**
Mocks [here](https://www.figma.com/file/fMQodZKzn9B5aZTN5G0fLJ/dYdX-%E2%80%BA-Desktop?type=design&node-id=27345-207&mode=design&t=MQWOIgzQVZCubM20-0).

For ease of implementation, we've only implemented notifications for:
- Created trigger order (successfully / failed)
- Cancelled trigger order (successfully / failed)

Note that this means replaced orders show a cancelled trigger order notification followed by a created trigger order.
- It is possible for either one or both of the above to fail^, in which case the notifications will show as creating/cancelling failing

| | Creating | Cancelling |
| ---- | ---- | ---- |
| Error | Screenshot 2024-04-12 at 11 19 46 AM | Screenshot 2024-04-12 at 11 19 33 AM |
| Success | Screenshot 2024-04-12 at 11 41 49 AM | Screenshot 2024-04-12 at 11 43 28 AM | 

Preferences Dialog
| Flag on | Flag off |
| ---- | ---- |
| Screenshot 2024-04-12 at 11 52 29 AM | Screenshot 2024-04-12 at 11 52 52 AM |

*tested error states by updating local abacus to always send back an error and building

https://github.com/dydxprotocol/v4-web/assets/70078372/95b17a6d-0aa2-4e2a-9bf7-b04f3dd4a138

https://github.com/dydxprotocol/v4-web/assets/70078372/b9ff2ab8-fd83-4a11-8073-29af8cce7124

https://github.com/dydxprotocol/v4-web/assets/70078372/83b30a2b-db46-446b-bc2e-1c8155feb88b

https://github.com/dydxprotocol/v4-web/assets/70078372/1c29cbbe-3f28-4a76-8e14-43181e040d07

---

## Views

* New: `views/notifications/TriggerOrderNotification/index.tsx`
  * TriggerOrderNotification component, note that in our current state (which does not support order replacement on stateful orders), we only handle showing notifications for canceling (removing) or creating orders. We *do not* handle showing notifications for updating orders (an updated order shows as a cancel, followed by a replace). Reasoning here is that in the backend, this is how it actually functions, and with our current payload return pattern, it's v hairy to actually determine whether it was a replacement
* `src/views/dialogs/PreferencesDialog.tsx`
  * Update to render SL/TP notification preference if SL/TP flag is on (exact localization still TBD, follow up linked)
* `TriggersForm`
  * Updated to trigger notifications on async `placeTriggerOrders` call
  * We have to look at the component's current state (of stop loss + take profit orders that are being rendered) to extract all the correct info for the notifications since the payload itself does not give us enough information (specifically in the case of the `CancelOrderPayload)
  * Sets submit button to loading state until we recieve back our first error / success from abacus
  
## Components

* `src/views/OrderStatusIcon.tsx`
  * Updated to make `totalFilled` optional since it's only really used in one case

## Constants/Types

* `constants/localStorage.ts`
  * added key for `TriggerOrderNotifications`
* `constants/notifications.ts`
  * added `TriggerOrderNotification` notification type 

## Functions

* new: `lib/notifications.ts`
  * added a big helper function that determines the title + body text for a trigger order notification depending on the status, type, orderType

## Hooks

* `hooks/useLocalNotifications`
  * added `allTriggerNotifications` to state, in the same way we have `allTransferNotifications`
  * expose `triggerOrderNotifications`, `addTriggerOrderNotification`
* `hooks/useNotificationTypes`
  * Update to handle the `NotificationType.TriggerOrder` type, renders a `TriggerOrderNotification`
* `hooks/useNotifications`
  * Enable trigger order notifications by default
* `hooks/useSubaccount`
  * Update `placeTriggerOrders. onError` callback to have all the info we need to render the notification
  
## Packages

* `v4-localization` v1.1.61 -> v1.1.63

---

### CODE REVIEW

File: src/constants/localStorage.ts [hunk 1]
```diff
@@ -13,6 +13,7 @@ export enum LocalStorageKey {
   PushNotificationsEnabled = 'dydx.PushNotificationsEnabled',
   PushNotificationsLastUpdated = 'dydx.PushNotificationsLastUpdated',
   TransferNotifications = 'dydx.TransferNotifications',
+  TriggerOrderNotifications = 'dydx.TriggerOrderNotifications',
```
Comment (aforaleka) [inline src/constants/localStorage.ts:16]: mm do we need to store these in localstorage? (do they need to persist across sessions?)

Comment (moo-onthelawn) [reply src/constants/localStorage.ts:16]: hmm, it's probably not important for them to persist 🤔 lemme double check with product

Comment (moo-onthelawn) [reply src/constants/localStorage.ts:16]: hmm actually confirmed with design that we do want them to persist (would want users to see it in their notif bar even if they accidentally closed it). We could set an expiration date for them, but that's not currently supported by the framework so will probably defer looking into that unless feedback comes in

Comment (aforaleka) [reply src/constants/localStorage.ts:16]: i meant across sessions as in after user closes the window! i don't think we need to save those notifications right? they can probably still be in the notification bar without saving to local storage?

Comment (moo-onthelawn) [reply src/constants/localStorage.ts:16]: hmm so the notification bar unfortunately does seem to only read from local storage (it calls [useNotifications](https://github.com/dydxprotocol/v4-web/blob/main/src/views/menus/NotificationsMenu.tsx#L50) which gets its notifications from [localStorage](https://github.com/dydxprotocol/v4-web/blob/a8f976c62cf987c23ac23d43a0b2336ab20cedd2/src/hooks/useNotifications.tsx#L49) (and slowly filters them down)). Are you concerned about the pop ups coming up after the user opens a new window if they haven't dismissed them?

File: src/hooks/useLocalNotifications.tsx [hunk 2]
```diff
@@ -38,6 +37,16 @@ const useLocalNotificationsContext = () => {
     },
   });
 
+  const [allTriggerOrderNotifications, setAllTriggerOrderNotifications] = useLocalStorage<{
```
[the remark this answers is not in the document]
Comment (moo-onthelawn) [reply src/hooks/useLocalNotifications.tsx:40]: ooh good point - filed follow up [here](https://linear.app/dydx/issue/CT-771/clear-local-notifications-fifo-probably-when-cache-gets-too-large)

**Pull request merged**